Sources/Datadog/DatadogConfiguration.swift

```diff
@@ -31,44 +31,126 @@ extension Datadog {
                 return Configuration(
                     clientToken: clientToken,
-                    logsEndpoint: logsEndpoint
+                    logsEndpoint: logsEndpoint,
+                    serviceName: serviceName,
+                    environment: environment
                 )
             }
         }
     }
+
+    /// Valid SDK configuration, passed to the features.
+    ///
+    /// It takes two types received from the user: `Datadog.Configuration` and `AppContext` and blends them together
+    /// with resolving defaults and ensuring the configuration consistency.
+    internal struct ValidConfiguration {
+        internal let applicationName: String
+        internal let applicationVersion: String
+        internal let applicationBundleIdentifier: String
+        internal let serviceName: String
+        internal let environment: String
+
+        internal let logsUploadURLWithClientToken: URL
+
+        init(configuration: Datadog.Configuration, appContext: AppContext) throws {
+            self.init(
+                applicationName: appContext.bundleName ?? appContext.bundleType.rawValue,
+                applicationVersion: appContext.bundleVersion ?? "0.0.0",
+                applicationBundleIdentifier: appContext.bundleIdentifier ?? "unknown",
+                serviceName: configuration.serviceName ?? appContext.bundleIdentifier ?? "ios",
+                environment: try ifValid(environment: configuration.environment),
+                logsUploadURLWithClientToken: try ifValid(
+                    endpointURL: configuration.logsEndpoint.url,
+                    clientToken: configuration.clientToken
+                )
+            )
+        }
+
+        init(
+            applicationName: String,
+            applicationVersion: String,
+            applicationBundleIdentifier: String,
+            serviceName: String,
+            environment: String,
+            logsUploadURLWithClientToken: URL
+        ) {
+            self.applicationName = applicationName
+            self.applicationVersion = applicationVersion
+            self.applicationBundleIdentifier = applicationBundleIdentifier
+            self.serviceName = serviceName
+            self.environment = environment
+            self.logsUploadURLWithClientToken = logsUploadURLWithClientToken
+        }
+    }
+}
+
+private func ifValid(environment: String) throws -> String {
+    let regex = #"^[a-zA-Z0-9_]+$"#
+    guard environment.range(of: regex, options: .regularExpression, range: nil, locale: nil) != nil else {
+        throw ProgrammerError(description: "`environment` contains illegal characters (only alphanumerics and `_` are allowed)")
+    }
+    return environment
+}
+
+private func ifValid(endpointURL: String, clientToken: String) throws -> URL {
+    guard !endpointURL.isEmpty, let endpointURL = URL(string: endpointURL) else {
+        throw ProgrammerError(description: "`endpointURL` cannot be empty.")
+    }
```

Comment:
`let endpointURL = URL(string: endpointURL)`
could we use a different name when we change the type?

---

> "`endpointURL` cannot be empty."

`endpointURL` function parameter could be non-nil yet invalid so that `ProgrammerError` with this message would be thrown although `endpointURL` was not empty

Reply:
👌 Changed to:
```swift
guard let endpointURL = URL(string: endpointURLString) else { ... }
```
which covers 🎸 both things.